Removing env-host-type in preparation of env and environment changes.

kwoodson opened the pull request.

detiber commented on playbooks/gce/openshift-cluster/config.yml:
```diff
@@ -15,10 +15,10 @@
+    g_etcd_group: "{{ groups['tag_host-type-openshift-etcd'] | intersect(groups['tag_environment_' ~ cluster_id}}"
+    g_lb_group: "{{ 'tag_host-type-openshift-lb' | intersect(groups['tag_environment_' ~ cluster_id] }}"
+    g_masters_group: "{{ 'tag_host-type-openshift-master' | intersect(groups['tag_environment_' ~ cluster_id] }}"
+    g_nodes_group: "{{ 'tag_host-type-openshift-node' | intersect(groups['tag_environment_' ~ cluster_id] }}"
```
If we are going to change g_*_group to be a list of hosts instead of a group name, then I'd rather that we rename them to g_*_hosts.

As an aside, we'll need to update the other playbooks (byo, openstack, libvirt) as well for these changes, otherwise they will be broken with this change.

Another thing to consider is that with the gce support of tag_<tag-name>_<tag_val> group names, we could further consolidate the aws, gce, openstack and libvirt config playbooks to share a common cloud provider set of playbooks that can do things like set tags to be used, group names based on tags and other common variables. The libvirt provider playbooks would need to be updated to change the names to use the same tag_<tag-name>_<tag_val> format to reach that though.

kwoodson replied:
@detiber,  I have made the other changes to the other playbooks.

twiest replied:
I agree with @detiber, these vars should be renamed to g_*_hosts.

I really like the idea of having these be simple host lists as they don't seem to be being used as host groups in any playbooks (in this repo, anyway).

detiber commented:
I'm a bit skeptical of the need to change the `g_*_group` variables to be a list of hosts instead of a group name, since it changes the interaction with the common playbooks (and there *could* be users out there that leverage the common playbooks only).

I also think that any change to what we are tagging and how we are referencing groups should be done across all of our cloud providers (even if it is through a separate PR that is reviewed by end users of those providers). I don't think we should expect them to clean up and backport consistency changes to match the providers that we use (unless we start collecting maintainers for the different cloud provider scripts and reach out to them proactively when we make changes like this).

That said, I am all for reducing the use of combined tags.

kwoodson commented:
@twiest @menren @lhuard1A @detiber ,

I'm making a preliminary change to remove env_host-type tags.  I have changed as many locations as I could find.  This will prepare us and simplify the env and environment tag changes referenced in #958.

If you could review and test that would be great.

twiest commented on playbooks/adhoc/s3_registry/s3_registry.yml:
```diff
@@ -6,7 +6,7 @@
 # The 'clusterid' is the short name of your cluster.
 
-- hosts: tag_env-host-type_{{ clusterid }}-openshift-master
+- hosts: tag_environment_{{ env }}:&tag_host-type_openshift-master
```
I would think that this would still use clusterid for now.

In other words, I think this should be:
```
- hosts: tag_environment_{{ clusterid }}:&tag_host-type_openshift-master
```

detiber commented:
@kwoodson The following playbooks will need to be updated for the `g_*_group` change:
- playbooks/byo/openshift-cluster/config.yml
- playbooks/byo/openshift-cluster/scaleup.yml
- openshift-cluster/upgrades/v3_0_minor/upgrade.yml
- openshift-cluster/upgrades/v3_0_to_v3_1/upgrade.yml
They all currently reference lines like this:
```
    g_masters_group: "{{ 'masters' }}"
```
which will now need to be:
```
    g_masters_group: "{{ groups.masters }}"
```

Similarly the following playbooks will need to be adjusted as well:
- playbooks/aws/openshift-cluster/scaleup.yml
- playbooks/aws/openshift-cluster/upgrades/v3_0_to_v3_1/upgrade.yml

twiest commented on playbooks/common/openshift-cluster/evaluate_groups.yml:
```diff
@@ -25,7 +25,7 @@
       ansible_ssh_user: "{{ g_ssh_user | default(omit) }}"
       ansible_sudo: "{{ g_sudo | default(omit) }}"
-    with_items: groups[g_etcd_group] | default([])
+    with_items: "{{ g_etcd_group | default([])}}"
```
minor nit, it needs a space at the end:
```
with_items: "{{ g_etcd_group | default([]) }}"
```

kwoodson commented:
@detiber,

I didn't see playbooks/aws/openshift-cluster/upgrades/v3_0_to_v3_1/upgrade.yml.  Is that in master?

Other I have changed as per your recommendation.

detiber commented:
```
> find . -name upgrade.yml
./playbooks/aws/openshift-cluster/upgrades/v3_0_to_v3_1/upgrade.yml
./playbooks/common/openshift-cluster/upgrades/v3_0_to_v3_1/upgrade.yml
./playbooks/common/openshift-cluster/upgrades/v3_0_minor/upgrade.yml
./playbooks/byo/openshift-cluster/upgrades/v3_0_to_v3_1/upgrade.yml
./playbooks/byo/openshift-cluster/upgrades/v3_0_minor/upgrade.yml

> git log playbooks/aws/openshift-cluster/upgrades/v3_0_to_v3_1/upgrade.yml 
commit 91411addfbaec7db5a039358d9fac8736f064b57
Author: Stefanie Forrester <sedgar@redhat.com>
Date:   Wed Nov 25 13:03:58 2015 -0500

    added upgrade playbook for online

```

It was merged to master this morning I believe.

twiest commented on playbooks/gce/openshift-cluster/config.yml:
```diff
@@ -15,10 +15,10 @@
-    g_lb_group: "{{ 'tag_env-host-type-' ~ cluster_id ~ '-openshift-lb' }}"
-    g_masters_group: "{{ 'tag_env-host-type-' ~ cluster_id ~ '-openshift-master' }}"
-    g_nodes_group: "{{ 'tag_env-host-type-' ~ cluster_id ~ '-openshift-node' }}"
+    g_etcd_group: "{{ groups['tag_host-type-openshift-etcd'] | intersect(groups['tag_environment_' ~ cluster_id}}"
```
I just checked and GCE doesn't have tag_environment_, it has tag_env-.

For example: `tag_env-ops`

So, I think this should be:
```
    g_etcd_group: "{{ groups['tag_host-type-openshift-etcd'] | intersect(groups['tag_env-' ~ cluster_id }}"
```

Obviously we should fix this in the followup clusterid PR, but for this one, I think it should remain with what's there now (so it works).

twiest commented on playbooks/libvirt/openshift-cluster/config.yml:
```diff
@@ -14,10 +14,10 @@
-    g_lb_group: "{{ 'tag_env-host-type-' ~ cluster_id ~ '-openshift-lb' }}"
-    g_masters_group: "{{ 'tag_env-host-type-' ~ cluster_id ~ '-openshift-master' }}"
-    g_nodes_group: "{{ 'tag_env-host-type-' ~ cluster_id ~ '-openshift-node' }}"
+    g_etcd_group: "{{ groups['tag_host-type-openshift-etcd'] | intersect(groups['tag_environment_' ~ cluster_id]) }}"
```
Are you sure this has tag_environment_ ?

Like GCE, I suspect libvirt uses dashes, and it probably didn't get changed from env to environment.

twiest commented:
@kwoodson I really like the direction of this PR. Thanks for getting this cleanup effort going.

kwoodson commented:
@detiber,

I have updated the upgrade.yml files as you mentioned in your previous comment.

@twiest @detiber, I have renamed the 'g_masters_group' to 'g_master_hosts'.

detiber commented on playbooks/common/openshift-cluster/evaluate_groups.yml:
```diff
@@ -33,11 +33,11 @@
+  # Use g_new_node_hosts if it exists otherwise g_node_hosts
   - set_fact:
-      g_nodes_to_config: "{{ g_new_nodes_group | default(g_nodes_group | default([])) }}"
+      g_node_to_config: "{{ g_new_node_hosts | default(g_node_hosts | default([])) }}"
```
g_node_to_config should probably be g_nodes_to_config or g_node_hosts_to_config.

The current naming seems to imply a single host instead of multiple.

detiber commented:
@kwoodson looks good outside of my naming nit.

I haven't had a chance to test, but nothing is standing out to me.  We can probably enlist Lenaic to test libvirt and OpenStack for us.

kwoodson commented:
@detiber,  Thanks for the review.  I have recently updated the tags to the proper ones that actually work.

I ran this against AWS and it was successful.
localhost                  : ok=89   changed=9    unreachable=0    failed=0   
opstest-master-1c5f2       : ok=232  changed=66   unreachable=0    failed=0   
opstest-node-compute-f146d : ok=72   changed=27   unreachable=0    failed=0   
opstest-node-infra-9e0b4   : ok=72   changed=27   unreachable=0    failed=0   

@detiber, @twiest, @lhuard1A, @menren,  Would you mind reviewing and testing with this PR?  It passes the test on aws but I do not have a setup to test openstack, gce, or byo.

menren commented:
Hello,

Did an origin deployment on GCE. Failed with this output  :  
```
TASK: [add_host ] ************************************************************* 
ok: [localhost]

PLAY [Populate oo_hosts_to_update group] ************************************** 

TASK: [Evaluate oo_hosts_to_update] ******************************************* 
fatal: [localhost] => with_items expects a list or a set

FATAL: all hosts have already failed -- aborting

PLAY RECAP ******************************************************************** 
           to retry, use: --limit @/root/launch.retry

localhost                  : ok=23   changed=4    unreachable=0    failed=1   

```

kwoodson commented:
@menren, investigating

I believe I have fixed this for gce.  Please retest.

detiber commented:
@abutcher can you verify this with the byo playbooks?

abutcher commented:
Failing for byo with:

```
TASK: [Evaluate oo_lb_to_config] **********************************************
fatal: [localhost] => with_items expects a list or a set

FATAL: all hosts have already failed -- aborting
```

kwoodson commented:
@abutcher, I tried to update those to "{{  groups.lb | default([]) }}".  Could you retest?

abutcher commented:
@kwoodson That works. Looks like there are a few other places where we'll need the defaults, like upgrade and scaleup.

kwoodson commented:
@abutcher, added defaults to those files.

abutcher commented:
LGTM

menren commented:
I ran again on GCE : 
```
PLAY [Populate oo_hosts_to_update group] **************************************

TASK: [Evaluate oo_hosts_to_update] ******************************************* 
fatal: [localhost] => Failed to template {{ g_master_hosts | union(g_node_hosts) | union(g_etcd_hosts) | default([]) }}: an unexpected type error occurred. Error was coercing to Unicode: need string or buffer, list found

FATAL: all hosts have already failed -- aborting

```

lhuard1A commented:
I have the same error as @menren on **libvirt**:

```
$ bin/cluster create -t enterprise -o image_name=rhel-guest-image-7.2-20151102.0.x86_64.qcow2 libvirt lenaic
```
```
PLAY [Populate oo_hosts_to_update group] ************************************** 

TASK: [Evaluate oo_hosts_to_update] ******************************************* 
fatal: [localhost] => Failed to template {{ g_master_hosts  | union(g_node_hosts) | union(g_etcd_hosts) | default([]) }}: an unexpected type error occurred. Error was coercing to Unicode: need string or buffer, list found

FATAL: all hosts have already failed -- aborting
```

kwoodson commented:
@lhuard1A, Please test openstack again.  I have updated the group names.

@menren, I have run a test on gce locally and it worked.  Could you please retest?

lhuard1A commented:
@kwoodson : I have tested 35e3481 and it is working on both OpenStack and libvirt.

kwoodson commented:
@lhuard1A, thanks.

@menren, Would you mind running a quick test?  I was able to run this on gce but I wanted you to weigh in.

menren commented:
Test running

menren commented:
This is going further on GCE (origin deployment) but still fail on error : 
```
PLAY [Compute haproxy_backend_servers] **************************************** 

TASK: [set_fact ] ************************************************************* 
fatal: [localhost] => Failed to template {{ hostvars | oo_select_keys(groups['oo_masters_to_config']) | oo_haproxy_backend_masters }}: |failed expects first param is a list

FATAL: all hosts have already failed -- aborting
```

kwoodson commented:
@menren, looking

kwoodson commented:
@menren, Could you send me some output with bin/cluster -vv? I updated join_node.yml with the group variables. Could you also retest?

menren commented:
I launched it before leaving work, I have a quick look to see if I have the logs, otherwise I will launch again

menren commented:
Here is the log of bin/cluster create -vv gce
[gce.txt](https://github.com/openshift/openshift-ansible/files/59719/gce.txt)

kwoodson commented:
@menren,

It looks like the error is:
{{ hostvars | oo_select_keys(groups['oo_masters_to_config']) | oo_haproxy_backend_masters }}: |failed expects first param is a list

This error comes from oo_select_keys. My guess is that oo_masters_to_config is undefined.  It gets defined here:
https://github.com/kwoodson/openshift-ansible/blob/remove_env_host_type/playbooks/common/openshift-cluster/evaluate_groups.yml#L30-L36

evaluate_groups.yml gets called from:
* common/openshift-cluster/config.yml 
* common/openshift-cluster/scaleup.yml.  

These get called from 
* gce/openshift-cluster/update.yml
* gce/openshift-cluster/join_node.yml
* gce/openshift-cluster/config.yml

Each of these playbooks has the necessary g_master_hosts being defined.

I have updated since your last build.  Rerun again please.  I'm running out of ideas.

menren commented:
Ok I run again

menren commented:
With your last code : 
```
TASK: [openshift_master | fail ] ********************************************** 
failed: [github-master-591b8] => {"failed": true}
msg: openshift_master_cluster_method must be set to either 'native' or 'pacemaker' for multi-master installations

FATAL: all hosts have already failed -- aborting
```

command line used : ```bin/cluster -vv create gce -i 0 -n 2```

by default it's spawning 2 masters, is there any configuration to do for multi-master ? never tried it

[gce2.txt](https://github.com/openshift/openshift-ansible/files/59797/gce2.txt)

I try with mono-master

menren commented:
Good for mono-master deployment.

abutcher commented:
@menren BYO has multi-master but we haven't taken that into GCE/AWS/etc

https://github.com/openshift/openshift-ansible/blob/master/inventory/byo/hosts.ose.example#L70-L89

abutcher commented:
@menren Also the default value for master count in `bin/cluster` is 1 so I'm curious how 2 got set.

menren commented:
Maybe I had a residual master vm ... I thought I cleaned before deployment.
Don't have time to check now, cause it's late and it's weekend !
But I understand better the error because some of variables defined in BYO mode are probably not defined in case of GCE

Anyways for 1 master, it seems good to me now :+1:
